lib/dao.js

```diff
@@ -624,7 +622,208 @@ DataAccessObject.upsert = function(data, options, cb) {
   }
   return cb.promise;
 };
+/**
+ * Performs upsert based on the instance that matches the where clause
+ . This is a PUT operation and based on
+ * non-primary key. The filter takes one key value pair and updates single matching instance.*
```

Comment:
>  1) Do a retrieve operation based on a non-primary key and update, if there is a single matching instance, throw error, if multiple instances are fetched and create if there are no instances found.

+1 from me

> Still I don't see why upsertWithWhere method should not work with PK! It should still work with PK, although it would be the same functionality as upsert in this case.

I agree with @Amir-61 here. Personally, I think we should allow any `where` filter and not limit this operation to a single key/value condition. That's what I expect from API method called `upsertWithWhere`.

Now it looks to me that perhaps you need a different method @mountain1234585, something like `upsertByNonPrimaryKey(key, value, data)`. In my opinion, such method is too specialised to be added to LoopBack. However, once we have more generic `upsertWithWhere` available, `upsertByNonPrimaryKey` can be easily implemented in user-land, e.g. as a mixin.

```js
function mixinUpsertByNonPrimaryKey(Model) {
  Model.upsertByNonPrimaryKey = function(key, value, data, cb) {
    // check arguments are valid, apply any PK/non-PK limitations, etc.
    var where = {};
    where[key] = value;
    data[key] = value;
    this.upsertWithWhere(where, data, cb);
  };

  Model.setup = function() {
    Model.base.setup.apply(this, arguments);
    Model.remoteMethod('upsertByNonPrimaryKey', { /*... */ });
  };
}
```